Switching between Safari and other apps can result in gray backgrounds unexpectedly

mwyrzykowski opened the pull request:
#### bb983de430ea094f7eca3c4c10b36ab99ad46601
<pre>
Switching between Safari and other apps can result in gray backgrounds unexpectedly
<a href="https://bugs.webkit.org/show_bug.cgi?id=291534">https://bugs.webkit.org/show_bug.cgi?id=291534</a>
<a href="https://rdar.apple.com/149226548">rdar://149226548</a>

Reviewed by Gerald Squelart.

-[CALayer setPreferredDynamicRange:] does not work for non-accelerated
drawing, skip calling into the code path when the setting is disabled.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::setDynamicRangeLimitRecursive):

Canonical link: <a href="https://commits.webkit.org/293713@main">https://commits.webkit.org/293713@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/12e103b95b27d363f2a55378de503eea33d7a152

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/99635 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/19285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/9546 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/104766 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50231 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/19574 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/27718 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75847 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32942 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/102642 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/14913 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89966 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56224 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/14713 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/49595 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/84676 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8037 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107125 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/26750 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/19529 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/84813 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27114 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86170 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84321 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21421 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28994 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/6700 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/20539 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/26690 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/31893 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/26505 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/29821 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28075 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

squelart commented on Source/WebKit/UIProcess/mac/WebViewImpl.mm:
```diff
@@ -2171,7 +2171,8 @@ static void setDynamicRangeLimit(CALayer* layer, PlatformDynamicRangeLimit platf
 {
     m_page->setShouldSuppressHDR(suppress);
-    setDynamicRangeLimit(m_rootLayer.get(), suppress ? PlatformDynamicRangeLimit::defaultWhenSuppressingHDR() : PlatformDynamicRangeLimit::noLimit(), true);
+    if (m_page->protectedPreferences()->acceleratedDrawingEnabled())
+        setDynamicRangeLimit(m_rootLayer.get(), suppress ? PlatformDynamicRangeLimit::defaultWhenSuppressingHDR() : PlatformDynamicRangeLimit::noLimit(), true);
```
Since this is temporary bug-fixing, and we'll eventually need to get this working properly, would it be possible to file a follow-up to https://bugs.webkit.org/show_bug.cgi?id=291534, and add a `// FIXME` comment here with the new bug number?

mwyrzykowski replied:
That is a good idea, though I wonder if we need to support the `!acceleratedDrawingEnabled()` path? It's an internal setting and otherwise always 1 in the UIProcess afaik

squelart replied:
Ideally we'd want the behavior to stay the same between the two modes, but yeah if that's just an internal flag mostly used for debugging, it could potentially stay unsupported.
That's why I said "optionally" ;-) I'll let you decide.
Thank you for the quick patch.

squelart reviewed (approved):
r+, optionally with a FIXME pointing at a follow-up bug.

mwyrzykowski committed:
Switching between Safari and other apps can result in gray backgrounds unexpectedly
https://bugs.webkit.org/show_bug.cgi?id=291534
rdar://149226548

Reviewed by Gerald Squelart.

-[CALayer setPreferredDynamicRange:] does not work for non-accelerated
drawing, skip calling into the code path when the setting is disabled.

* Source/WebKit/UIProcess/mac/WebViewImpl.mm:
(WebKit::setDynamicRangeLimitRecursive):

Canonical link: https://commits.webkit.org/293713@main
